src/plans/solver_state.jl

```diff
@@ -207,67 +244,201 @@ iteration, i.e. acts on `(p,o,i)`, where `p` is a [`AbstractManoptProblem`](@ref
+tangent vector storage tu `TTS`. For example you may call
+`StorageStateAction(M, Symbol[], Tuple{:Iterate}, Tuple{:Gradient})` to create efficient
+storage for point representing iterate and tangent vector representing gradient.
+
+    function StoreStateAction(
+        general_keys::Vector{Symbol}=Symbol[],
+        point_values::NamedTuple=NamedTuple(),
+        tangent_values::NamedTuple=NamedTuple(),
+        once::Bool=true,
+    ))
 
 Initialize the Functor to a set of keys, where the dictionary is initialized to
 be empty. Further, `once` determines whether more that one update per iteration
 are effective, otherwise only the first update is stored, all others are ignored.
-"""
-mutable struct StoreStateAction <: AbstractStateAction
-    values::Dict{Symbol,<:Any}
-    keys::NTuple{N,Symbol} where {N}
+Make a copy of points and tangent vectors passed to `point_values` and `tangent_values`
+for later storage respective fields.
+"""
+mutable struct StoreStateAction{
+    TPS<:NamedTuple,TXS<:NamedTuple,TPI<:NamedTuple,TTI<:NamedTuple
+} <: AbstractStateAction
+    values::Dict{Symbol,Any}
+    keys::Vector{Symbol} # for values
+    point_values::TPS
+    tangent_values::TXS
+    point_init::TPI
+    tangent_init::TTI
     once::Bool
     last_stored::Int
     function StoreStateAction(
-        keys::NTuple{N,Symbol} where {N}=NTuple{0,Symbol}(), once=true
+        general_keys::Vector{Symbol}=Symbol[],
+        point_values::NamedTuple=NamedTuple(),
+        tangent_values::NamedTuple=NamedTuple(),
+        once::Bool=true,
     )
-        return new(Dict{Symbol,Any}(), keys, once, -1)
+        point_init = NamedTuple{keys(point_values)}(map(u -> false, keys(point_values)))
+        tangent_init = NamedTuple{keys(tangent_values)}(
+            map(u -> false, keys(tangent_values))
+        )
+        point_values_copy = NamedTuple{keys(point_values)}(
+            map(u -> copy(point_values[u]), keys(point_values))
+        )
+        tangent_values_copy = NamedTuple{keys(tangent_values)}(
+            map(u -> copy(tangent_values[u]), keys(tangent_values))
+        )
+        return new{
+            typeof(point_values),
+            typeof(tangent_values),
+            typeof(point_init),
+            typeof(tangent_init),
+        }(
+            Dict{Symbol,Any}(),
+            general_keys,
+            point_values_copy,
+            tangent_values_copy,
+            point_init,
+            tangent_init,
+            once,
+            -1,
+        )
     end
 end
+@inline function StoreStateAction(
+    M::AbstractManifold,
+    dictionary_symbols::Vector{Symbol},
+    ::Type{TPS},
+    ::Type{TTS};
+    p_init=rand(M),
+    X_init=zero_vector(M, p_init),
+    once=true,
+) where {TPS<:Tuple,TTS<:Tuple}
+    TPS_tuple = Tuple(TPS.parameters)
+    TTS_tuple = Tuple(TTS.parameters)
+    point_values = NamedTuple{TPS_tuple}(map(_ -> p_init, TPS_tuple))
+    tangent_values = NamedTuple{TTS_tuple}(map(_ -> X_init, TTS_tuple))
```

Comment:
I statt correct that this does not copy? Would that mean that all fields share the same point memory? That would not be so good.

Reply:
Hm and the new constructor still requires to pass `store_points=Tuple{:Iterate})` ? I was hoping wee could unify all 3 to an array of symbols?

Reply:
...and when I have understood it I would love to provide docs how to use this ;)

Reply:
finally: StoreStateAction is one of the few things where the manifold is now mandatory. can we fid a way around that? Without a manifold of course only the slow storage works.

Reply:
`copy`ing is currently done in the inner constructor, though that will have to change a bit.

> Hm and the new constructor still requires to pass `store_points=Tuple{:Iterate})` ? I was hoping wee could unify all 3 to an array of symbols?

`Tuple{:Iterate}`  makes it much easier to optimize. A vector of symbols is definitely an option, though it makes optimization more difficult and adds some unavoidable overhead -- that's why I'd prefer to use `Tuple{:Iterate}`.

I'm not sure what part of usage is unclear though? Apart, of course, from "why does it need to be like that".

Reply:
I just mean the constructor, not the inner stuff, but for now...I posted this 45 minutes ago and the number of new storage I was able to do including testing is still 0 – maybe I am the stupid here, but if I as a developer of this package, can not use that properly, it might be the case for others as well.

Reply:
> finally: StoreStateAction is one of the few things where the manifold is now mandatory. can we fid a way around that? Without a manifold of course only the slow storage works.

In the current inner constructor you don't need to provide a manifold. You can just use the same old syntax (just with a vector instead of a tuple). Would you prefer a different argument list?

Reply:
> "why does it need to be like that".

It‘s not the why it's the how. I am mainly guessing and sometimes it works and sometimes not and I am not sure why. Currently my tests that I start writing randomly fail (probably not that random but) without me understanding why.

Reply:
There might be a bug in my code, so maybe I could help if you posted how you make a storage object and what errors you get?

Reply:
So I got a first one working but I was fighting for a bit with whether its

* `store_points = [:Iterate, :SecondPoint]` (my preferred way
* `store_points = Tuple(:Iterate, :SecondPoint)` my second idea what it might be
* `store_points = typeof(Tuple(:Iterate, :SecondPoint))` (saw that somewhere)
* `store_points = Tuple{:Iterate, :SecondPoint}` <- I think this is the right one?

or is there another `typeof` in between?

That is maybe the summary of my problem with the current constructor, it is not yet documented and does not provide an intuitive way (which in my opinion should be the first). And it is a constructor, so I do not care about performance. For example currently the symbols are extracted from `TTP.parameters` why not directly allow a vector of symbols?

Reply:
Sorry, I forgot to document it. It goes like this:
```julia
StorageStateAction(M; store_fields=Symbol[], store_points=Tuple{:Iterate}, store_vectors=Tuple{:Gradient})
```

Reply:
> And it is a constructor, so I do not care about performance. For example currently the symbols are extracted from `TTP.parameters` why not directly allow a vector of symbols?

It's... complicated. Mostly: it's me doing what I know works instead of spending hours benchmarking a bit nicer alternative in various scenarios making sure there are no big performance problems. Interactions of type unstable code (and with `[:some_point]` instead of `Tuple{:some_point}` we have type unstable code) with Julia's type manipulations have caused me many problems in the past.

Reply:
Hm, but now I use hours trying to guess how to use it. And I do not feel that this is really usable. So for best of cases no one else will touch these internals, if they would like to, we might loose that user since that is far away from an intuitive interface design.

Reply:
I will modify the constructor to make it accept both tuple types and vectors of symbols. It can easily accept either with no performance problems so maybe that would be the best solution?

Reply:
What we definitely need is updated docs for the StateActoinStorage, but I would really prefer if that was more unified, I am still nowhere in the storage sure which parameters to use from the list above (ok the first one never uses, but thats the one I would like to use).

Reply:
Now you can use it like `StoreStateAction(M; store_fields=[:my, :fields, :to, :store], store_points=[:Iterate], store_vectors=[:Gradient])` (fast) or  `StoreStateAction([:my, :fields, :to, :store, :Iterate, :Gradient])` (slow). I don't think it can get any more unified?

Reply:
That would have spared me an hour today having that ;) I will check it just now and write the docs.

Reply:
I love the new interface (but I am lazy for now and since the previous one works as well leave the current code as is).